Mathlib/CategoryTheory/DifferentialObject.lean

```diff
@@ -0,0 +1,368 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+
+! This file was ported from Lean 3 source module category_theory.differential_object
+! leanprover-community/mathlib commit 6876fa15e3158ff3e4a4e2af1fb6e1945c6e8803
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Int.Basic
+import Mathlib.CategoryTheory.Shift.Basic
+import Mathlib.CategoryTheory.ConcreteCategory.Basic
+
+/-!
+# Differential objects in a category.
+
+A differential object in a category with zero morphisms and a shift is
+an object `X` equipped with
+a morphism `d : X ⟶ X⟦1⟧`, such that `d^2 = 0`.
+
+We build the category of differential objects, and some basic constructions
+such as the forgetful functor, zero morphisms and zero objects, and the shift functor
+on differential objects.
+-/
+
+
+open CategoryTheory.Limits
+
+universe v u
+
+namespace CategoryTheory
+
+variable (C : Type u) [Category.{v} C]
+
+-- TODO: generalize to `HasShift C A` for an arbitrary `[AddMonoid A]` `[One A]`.
+variable [HasZeroMorphisms C] [HasShift C ℤ]
+
+/-- A differential object in a category with zero morphisms and a shift is
+an object `pt` equipped with
+a morphism `d : X ⟶ X⟦1⟧`, such that `d^2 = 0`. -/
+-- Porting note: Removed `@[nolint has_nonempty_instance]`
+structure DifferentialObject where
+  /-- The underlying object of a differential object. -/
+  pt : C
```

Comment:
I don't think `pt` is a good name. It was okay for (co)cones, but doesn't seem to make sense here. How about `obj`?